Java 17 Support with Sigtest 2.2 from JakartaEE TCK Tools

lkishalmi opened the pull request:
I've tried to upgrade Sigtest to the latest available version 1.7, and that still fails on Java 17.
@mbien pionted out that there is a fork in JakartaEE TCK that could work. It did the job.

This is needed when a module is on Java 17 and exposes some API.

lkishalmi commented:
Well, this version seems to fail on `plarfom/libs.osgi` which shall generate an empty report.
The new tool has some specific handling for Java 17.

The old tool does something somehow not match any package with the following patterns `info.dmtree.**, org.osgi.**`

The new version does. Unfortunately `osgi.cmpn-7.0.0.jar` has references some `javax.servlet` classes which are not on the classpath.

Adding `sigtest.public.packages=-` to the `project.properties` helped, but there are other classloader issues after that.

I'm out of ideas here.

@jtulach ?

matthiasblaesing commented:
Not sure if it helps, but I think this causes the "error" in `libs.osgi`:

https://github.com/eclipse-ee4j/jakartaee-tck-tools/commit/56e3e338dba7b008536f5aa13d4423179d1d4b17

If I read the summary of the commit correctly, before this commit recursive definitions like `info.dmtree.**, org.osgi.**` were not handled correctly. If I'm not mistaken, only the first level was considered. 

The commit was added between 1.5 and 1.6 and indeed using 1.5 I can run the sigtest generator in `libs.osgi` and 1.6 fails. And for some unknown reason that also matches the fact, that NetBeans itself stayed with sigtest 1.4.

Channeling my bash deamons I came up with this list of modules that might be affected:

```
find . -name project.xml | xargs grep -l subpackages | cut --delimiter "/" -f 2-3
extide/libs.gradle
webcommon/libs.nashorn
java/j2ee.eclipselink
java/maven.embedder
enterprise/web.jsf12
enterprise/web.jsf20
enterprise/j2eeapis
ide/servletapi
ide/libs.lucene
ide/libs.jcodings
ide/o.apache.xml.resolver
ide/libs.snakeyaml_engine
ide/libs.xerces
ide/libs.flexmark
platform/libs.asm
platform/libs.osgi
```

lkishalmi commented:
Well the last commit forces to use sigtest to generate signatures without recursion.

That behavior was fixed in the commit mentioned above, so that commit is just keeping the same wrong way of doing the things as before.

mbien commented on enterprise/j2eeapis/nbproject/project.properties:
```diff
@@ -17,3 +17,5 @@
 is.autoload=true
 release.external/deployment-api-1.2-rev-1.jar=modules/ext/jsr88javax.jar
+
+sigtest.public.packages=javax.enterprise.deploy
```
this restricts the sig test to only the listed package prefix?

lkishalmi replied:
Yes. That's the idea, as the recursive behavior was not working prior 1.6. They checked only the listed part of the package hierarchy. So adding these mimics the former behavior.

mbien reviewed (commented):
thanks for updating this dependency! Its interesting that it went [full circle,](https://github.com/eclipse-ee4j/jakartaee-tck-tools/tree/master/tools/sigtest#sigtest) I believe the jakarta lib is a fork of the netbeans fork - and we would use it now again ;)

tests are green, couldn't find any regressions while browsing through the log (the familiar `java.lang.ClassFormatError: Index out of the constant pool bounds` traces are still there)

+1 from me, but someone who has more experience with the sig tests and public API requirements should take a look too.

ebarboni commented:
LGTM with my limited understanding

lkishalmi commented:
Anyone, a word? I'm about to merge this one today or tomorrow, as #7024 is depending on this one.

matthiasblaesing commented on extide/libs.gradle/nbproject/project.properties:
```diff
@@ -21,5 +21,6 @@ javac.compilerargs=-Xlint -Xlint:-serial
 
 # For more information, please see http://wiki.netbeans.org/SignatureTest
 sigtest.gen.fail.on.error=false
+sigtest.public.packages=org.gradle
```
This is wrong - and masks the real problem: sigtest fails to parse the class files. IMHO we should in this case disable sigtest and not pretend it does something sensible here. A comment documenting the problem might help future readers.

matthiasblaesing commented on java/j2ee.eclipselink/nbproject/project.properties:
```diff
@@ -38,3 +38,5 @@ release.external/jakarta.persistence-2.2.3.jar=modules/ext/eclipselink/jakarta.p
 
 extra.module.files=modules/ext/docs/jakarta.persistence-2.2.3-doc.zip
+
+sigtest.public.packages=org.eclipse,javax.persistence
```
Strange error here when generating signature file.

lkishalmi replied:
Have to check it through, this one could be just used as another Ant library, in that case this shall not offer a public API.

matthiasblaesing reviewed (commented):
I would not go down the `sigtest.public.packages` property route. It is doomed to fail because in the future nobody will remember why on earth these broken declarations were made. In most cases I saw the packages would need to be declared as `package.**` (i.e. cover all subpackages) and then the original errors reappear.

I wrote it inline already and would suggest to recreate the sigfile where it works and where not disable sigtest with a comment explaining why it is deactivated.

lkishalmi commented:
@matthiasblaesing thanks for your feedback!

lkishalmi commented on enterprise/web.jsf20/nbproject/project.properties:
```diff
@@ -20,3 +20,5 @@ javac.source=1.8
 release.external/javax.faces-2.3.9-license.txt=modules/ext/jsf-2_2/license.txt
 spec.version.base=1.57.0
+
+sigtest.public.packages=javax.faces,com.sun.faces
```
Strangely, this did not occur to me.

Though I think these definitions `web.jsf*` are there for providing libraries to Ant based projects. So technically they should not provide any public api to any NB modules. Also JSF 1.2 is soon 18 years old. Probably it would be better not to include the jars in NB. 

web.jsf20  hide JSF 2.3 we seem to be built our JSF support on that. 

I'm going to file separate PR-s to remove the JSF 1.2  libraries.

lkishalmi commented on enterprise/web.jsf12/nbproject/project.properties:
```diff
@@ -17,3 +17,5 @@
 release.external/jsf-api-1.2_05.jar=modules/ext/jsf-1_2/jsf-api.jar
 spec.version.base=1.48.0
+
+sigtest.public.packages=javax.faces
```
Added `org.netbeans.modules.j2ee.dd` as well as `javax.annotations.Resource` is needed beyond java 8.

mbien commented:
would it make sense to open a meta issue afterwards which has a list of all mods where sig tests are turned off so that we can look that up when needed?

matthiasblaesing commented on ide/libs.flexmark/nbproject/project.properties:
```diff
@@ -15,7 +15,7 @@
 # under the License.
 
-javac.source=1.8
+javac.source=11
```
I think this can be reverted as only the sigtest is disabled.

lkishalmi replied:
Reverted.

matthiasblaesing commented on ide/libs.lucene/nbproject/project.properties:
```diff
@@ -19,3 +19,8 @@ is.autoload=true
+
+# This is an very old library, the complete dependencies were never explored.
+# Since subpackage-s are working in new sigtest version, generation of the
+# signature file fails. Disabling that now.
```
Disabling is fine, the comment though does not match the reason (from my POV). The reported error is:

```
/home/matthias/src/netbeans/nbbuild/templates/projectized.xml:765: 
java.lang.AssertionError: org.apache.lucene.search.CachingWrapperFilter$FilterCache<org.apache.lucene.search.DocIdSet>
	at com.sun.tdk.signaturetest.loaders.BinaryClassDescrLoader.load(BinaryClassDescrLoader.java:225)
	at com.sun.tdk.signaturetest.core.ClassHierarchyImpl.load(ClassHierarchyImpl.java:180)
	at com.sun.tdk.signaturetest.core.ClassHierarchyImpl.getClassInfo(ClassHierarchyImpl.java:292)
	at com.sun.tdk.signaturetest.core.ClassHierarchyImpl.isInterface(ClassHierarchyImpl.java:272)
	at com.sun.tdk.signaturetest.core.ClassCorrector.fixType(ClassCorrector.java:315)
	at com.sun.tdk.signaturetest.core.ClassCorrector.fixMethods(ClassCorrector.java:298)
```

So I would say this is the same as in other cases: sigtest fails to parse the API classes.

matthiasblaesing commented on java/j2ee.eclipselink/nbproject/project.properties:
```diff
@@ -38,3 +38,6 @@ release.external/jakarta.persistence-2.2.3.jar=modules/ext/eclipselink/jakarta.p
 
 extra.module.files=modules/ext/docs/jakarta.persistence-2.2.3-doc.zip
+
+# There are some unexplored dependencies (Java Mail, and JAX-RS API ?)
```
```suggestion
# There are some missing dependencies (Java Mail, and JAX-RS API ?)
```

I'm wondering whether it would make sense to disable sigtest here as the API definition is much to broad. ecpliselink has a huge set of `internal` packages, which I would not consider API.

matthiasblaesing commented on enterprise/web.jsf20/nbproject/project.properties:
```diff
@@ -20,3 +20,7 @@ javac.source=1.8
 release.external/javax.faces-2.3.9-license.txt=modules/ext/jsf-2_2/license.txt
 spec.version.base=1.57.0
+
+# Old library without fully explored dependencies better to disable sigtest
```
```suggestion
# Old library with too broadly defined API - sigtest would give more noise than benefit
# (legacy behavior of sigtest ignored subpackages)
```

matthiasblaesing reviewed (approved):
In general looks good. I don't agree with all comments, but not to problematic, I left a few inline suggestions.

lkishalmi committed:
Java 17 Support with Sigtest 2.2 from JakartaEE TCK Tools

Disabled sigtest where it does not work, fixed signature files/dependencies where it does.

Update enterprise/web.jsf20/nbproject/project.properties
Update java/j2ee.eclipselink/nbproject/project.properties
Co-authored-by: Matthias Bläsing <mblaesing@doppel-helix.eu>

jtulach commented:
> @jtulach ?

- @mbien just pointed me at this issue
- FYI: https://github.com/jtulach/netbeans-apitest/commit/5956873eb29da782cfab296f8e1493fc7eb58751
- e.g. everyone concluded that switching to [Jakarta EE TCK version](https://github.com/eclipse-ee4j/jakartaee-tck-tools/tree/master/tools/sigtest) is the best option
